optimum/habana/transformers/generation/utils.py

```diff
@@ -173,6 +186,92 @@ def _update_model_kwargs_for_generation(
 
         return model_kwargs
 
+    def _prepare_decoder_attention_mask(
+        self,
+        max_steps: int,  # current stopping criteria
+        batch_size: int,
+        pad_token_id: int,
+        device: str,
+        dtype: str = torch.int64,
+    ) -> torch.Tensor:
+        x = torch.zeros((batch_size, max_steps), device=device, dtype=dtype)
+        return x.index_fill(1, torch.tensor([0]), pad_token_id)  # First the position with pad_token_id
+
+    def _prepare_past_key_values(
+        self,
+        max_steps: int,  # current stopping criteria
+        batch_size: int,
+        device: str,
+    ) -> Tuple[torch.LongTensor, torch.LongTensor, Optional[torch.Tensor], Optional[torch.Tensor]]:
+        n_heads = self.config.decoder_attention_heads
+        embed_size_per_head = self.config.d_model // n_heads
+        key_states_ = torch.zeros(
+            (self.config.decoder_layers, batch_size, n_heads, max_steps, embed_size_per_head), device=device
+        )
+        value_states_ = torch.zeros(
+            (self.config.decoder_layers, batch_size, n_heads, max_steps, embed_size_per_head), device=device
+        )
+        past_key_values = tuple(
+            (key_states_[i, :, :, :, :], value_states_[i, :, :, :, :], None, None)
+            for i in range(self.config.decoder_layers)
+        )
+        return past_key_values
+
+    def _prepare_decoder_input_ids_for_generation(
+        self,
+        batch_size: int,
+        model_input_name: str,
+        decoder_start_token_id: int = None,
+        bos_token_id: int = None,
+        max_length: int = 1,
+        model_kwargs: Optional[Dict[str, torch.Tensor]] = None,
+        device: torch.device = None,
```

Comment:
Can you reorder the args in the same order as in Transformers please? https://github.com/huggingface/transformers/blob/v4.31.0/src/transformers/generation/utils.py#L648
And we can put `max_length` at the end.